File: rs/minsk/src/analysis/syntax/token.rs
use std::fmt::Display;

use crate::analysis::text_span::TextSpan;
use crate::object::Object;

use super::kind::SyntaxKind;
use super::node::SyntaxNode;

#[derive(Debug, Clone)]
pub struct SyntaxToken {
    pub(crate) kind: SyntaxKind,
    pub(crate) text: String,
    pub(crate) position: usize,
    pub(crate) length: usize,
    pub(crate) value: Option<Object>,
}

impl SyntaxToken {
    pub(crate) fn new(
        kind: SyntaxKind,
        text: String,
        position: usize,
        length: usize,
        value: Option<Object>,
    ) -> Self {
        Self {
            kind,
            text,
            position,
            length,
            value,
        }
    }

    pub(crate) fn span(&self) -> TextSpan {
        TextSpan::new(self.position, self.text.len())
    }

    pub(crate) fn position(&self) -> usize {
        self.position
    }

    pub(crate) fn length(&self) -> usize {
        self.length
    }

    pub(crate) fn value(&self) -> Option<&Object> {
        self.value.as_ref()
    }

    pub(crate) fn text(&self) -> &str {
        self.text.as_str()
    }
}

impl Display for SyntaxToken {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?} '{}'", self.kind, self.text)?;
        match self.value.as_ref() {
            Some(v) => write!(f, " {}", v)?,
            None => {}
        }
        Ok(())
    }
}

impl SyntaxNode for SyntaxToken {
    fn kind(&self) -> SyntaxKind {
        self.kind
    }

    fn children(&self) -> Vec<&dyn SyntaxNode> {
        vec![]
    }

    fn is_token(&self) -> bool {
        true
    }
}

File: py/minsk/analysis/syntax/token.py
from collections.abc import Iterable

from minsk.analysis.syntax.kind import SyntaxKind
from minsk.analysis.syntax.node import SyntaxNode
from minsk.analysis.text.span import TextSpan
from minsk.runtime.value import Value


class SyntaxToken(SyntaxNode):
    def __init__(
        self, kind: SyntaxKind, position: int, text: str, value: Value
    ) -> None:
        self._kind = kind
        self.position = position
        self.text = text
        self._value = value

    def __str__(self) -> str:
        result = f"{self.kind.name} '{self.text}'"
        if self.value is not None:
            result += f" {self.value}"
        return result

    @property
    def kind(self) -> SyntaxKind:
        return self._kind

    @property
    def span(self) -> TextSpan:
        return TextSpan(self.position, len(self.text))

    @property
    def children(self) -> Iterable[SyntaxNode]:
        return ()

    @staticmethod
    def is_token() -> bool:
        return True

    @property
    def value(self) -> Value:
        return self._value
